util/grub-editenv.c

```diff
@@ -143,29 +143,41 @@ typedef struct fs_envblk fs_envblk_t;
 fs_envblk_t *fs_envblk = NULL;
 
 #if defined(HAVE_LIBZFS) && defined(HAVE_LIBNVPAIR) && defined(HAVE_LIBZFS_BOOTENV)
-static void *
-grub_zfs_init (grub_device_t dev)
+static int
+grub_zfs_init (grub_device_t dev, void **out)
 {
   libzfs_handle_t *g_zfs = libzfs_init();
   int err;
   char *name;
   zpool_handle_t *zhp;
+  char buf[16];
 
   if (g_zfs == NULL)
-    return NULL;
+    return -1;
 
   err = fs_envblk->fs->label(dev, &name);
   if (err != GRUB_ERR_NONE) {
     libzfs_fini(g_zfs);
-    return NULL;
+    return -1;
   }
   zhp = zpool_open(g_zfs, name);
   if (zhp == NULL)
     {
       libzfs_fini(g_zfs);
-      return NULL;
+      return -1;
     }
-  return zhp;
+  /* If bootenv support isn't present on this system, don't try to use the envblk. */
+  err = zpool_get_bootenv (zhp, buf, 16, 0);
+  fprintf (stderr, "err: %d %d\n", err, libzfs_errno(zpool_get_handle(zhp)));
+  if (err == -1 && libzfs_errno(zpool_get_handle(zhp)) == EZFS_IOC_NOTSUPPORTED)
```

Comment:
Is it possible to get other errors that we should handle?

Reply:
Generally speaking, any other errors are going to prevent us from operating correctly. When we return an error, grub will print an error message, but there's not much else we can do about them.

Reply:
Most other errors have to do with specific issues to that invocation, or indicate that the envblock doesn't currently checksum. Either way, they don't necessarily stop us from working correctly.